Media: fix wrong paths in page view analytics

vindl opened the pull request:
Resolves: https://github.com/Automattic/wp-calypso/issues/23516

According to internal logs, some `calypso_page_view` paths coming from Media
were reported without replacing passed parameters with corresponding variables,
or removing wrong entries from the route. This was resolved by adding more
restrictive route matching for optional filter and site parameters.

### Testing instructions
1. Set `localStorage.setItem( 'debug', 'calypso:analytics:tracks' )` in the browser console to enable the debug.
2. To reproduce the issue, navigate to `/media/randomtext/{site}` or `/media/randomtext` and observe the `calypso_page_view` recorded paths in log.
3. Apply this branch and verify that `randomtext` is no longer accepted as valid media filter and recorded as such.
4. Check that there are no other valid filters that we should be matching here.
5. Try to break it with various paths that start with `/media/`.

vindl commented on client/my-sites/media/main.jsx:
```diff
@@ -268,10 +272,28 @@ class Media extends Component {
 		const site = this.props.selectedSite;
 		return (
 			<div ref="container" className="main main-column media" role="main">
+				<DocumentHead title={ translate( 'Media' ) } />
```
Note: I noticed that this is not setting the title correctly when switching filters. This is not good to merge until that is resolved.

gwwar replied:
From the old code wasn't this the case before? We can try and tidy that up in an additional janitorial if that's the case.

gwwar replied:
You're right there is an odd regression, where the page title goes back to the site name.

gwwar replied:
There's similar bad behavior with switching between comment filters. Maybe keep the old `setTitle` usage, and we can look into a separate PR for fixing DocumentHead

gwwar commented on client/my-sites/media/main.jsx:
```diff
@@ -336,4 +358,8 @@ class Media extends Component {
 
-export default localize( Media );
+const mapStateToProps = state => ( {
+	selectedSite: getSelectedSite( state ),
```
Doesn't hurt to also request `selectedSiteId` directly here:
```jsx
selectedSiteId: getSelectedSiteId( state ),
```

Copons replied:
Though, there is also a `selectedSite.slug` used, so, while it makes sense to directly select both, it'd also maybe clutter `mapStateToProps` too much?

vindl replied:
I avoided doing this because `selectedSite` object is passed to `EditorMediaModalDetail` and `MediaLibrary` components. I'd rather leave that refactor for a separate PR.

gwwar commented on client/my-sites/media/main.jsx:
```diff
@@ -268,10 +272,28 @@ class Media extends Component {
 	};
 
+	getAnalyticsPath = () => {
+		const { filter, selectedSite } = this.props;
```
Maybe let's invert this a bit, as a bonus it's a little easier to search for in an IDE:
```jsx
if ( selectedSiteId && filter ) { 
    //hmm what happens if the site isn't loaded yet? We pass an incorrect path to the page tracker component right?
    // a getSiteFragment check may be more appropriate
    return `/media/${ filter }/:site`
}

if ( selectedSiteId ) {
    return `/media/:site`;
}

return '/media';
```

vindl replied:
Afaik @Copons introduced a `hasSelectedSiteLoaded` check in `PageViewTracker` that should help guard against this case.

gwwar commented:
Looks close! I verified that normal page views still fire and that `http://calypso.localhost:3000/media/videos/blah/siteSlug` redirects as expected.
- [ ] Let's back out the page title changes until we can fix that in a separate PR
- [ ] Look into using getSiteFragment for `getAnalyticsPath` (I think we'll run into timing issues otherwise)
- [ ] Either in this PR or a follow up handle `http://calypso.localhost:3000/media/blah` and `http://calypso.localhost:3000/media/videos/blah` to redirect to site switcher. (Currently tries to render an empty all sites view)

Copons commented:
> Either in this PR or a follow up handle `http://calypso.localhost:3000/media/blah` and `http://calypso.localhost:3000/media/videos/blah` to redirect to site switcher. (Currently tries to render an empty all sites view)

I vote for tackling this here.
It can be as simple as adding
```es6
if ( ! getSiteFragment( context.path ) ) {
	return page.redirect( '/media' );
}
```
to the controller, or it could be improved with some more accurate `filter` handling (but it would also require updating the redirect logic in the index file).

Copons commented:
> Look into using getSiteFragment for `getAnalyticsPath` (I think we'll run into timing issues otherwise)

🤔
In theory, you just can't get into the `media/main.jsx` without a selected site (it would redirect to the sites selection or, well, open a blank screen 😄).
So, I'd question the entire value of checking `selectedSite` (or the site fragment) in `getAnalyticsPath`, because chances are it's always `true`.

vindl commented:
> Let's back out the page title changes until we can fix that in a separate PR

Reverted to old code.

> Look into using `getSiteFragment` for `getAnalyticsPath` (I think we'll run into timing issues otherwise)

I think @copons is right about not being able to reach this code without site being selected. I just removed that check.

> Either in this PR or a follow up handle http://calypso.localhost:3000/media/blah and http://calypso.localhost:3000/media/videos/blah to redirect to site switcher.

:+1: 

@gwwar all of the above updated in 04d564f2d3e57e9516be1481bbc80f20b4afa52b.

kwight commented:
Works great; redirects work well both with and without a site on the end. 👍
